src/HomeNet/Network/RoleServer.cs

```diff
@@ -612,54 +578,54 @@ public List<Client> GetClientListCopy()
     /// </summary>
     /// <param name="Data">Full ProtoBuf message to be processed.</param>
     /// <param name="Client">TCP client who send the message.</param>
-    public /*async*/ void ProcessMessageAsync(byte[] Data, Client Client)
+    public async Task ProcessMessageAsync(byte[] Data, Client Client)
     {
       string prefix = string.Format("{0}[{1}] ", this.logPrefix, Client.RemoteEndPoint);
       PrefixLogger log = new PrefixLogger(this.logName, prefix);
 
+      MessageBuilder messageBuilder = Client.MessageBuilder;
+
       log.Debug("()");
       try
       {
         // Update time until this client's connection is considered inactive.
         Client.NextKeepAliveTime = DateTime.UtcNow.AddSeconds(Client.KeepAliveIntervalSeconds);
-        log.Trace("Client ID 0x{0:X8} NextKeepAliveTime updated to {1}.", Client.Id, Client.NextKeepAliveTime.ToString("yyyy-MM-dd HH:mm:ss"));
+        log.Trace("Client ID 0x{0:X16} NextKeepAliveTime updated to {1}.", Client.Id, Client.NextKeepAliveTime.ToString("yyyy-MM-dd HH:mm:ss"));
 
-        Message message = Message.Parser.ParseFrom(Data);
-        log.Trace("Message type is {0}, message ID is {1}.", message.MessageTypeCase, message.Id);
-        switch (message.MessageTypeCase)
+        Message requestMessage = Message.Parser.ParseFrom(Data);
+        log.Trace("Received message type is {0}, message ID is {1}:\n{2}", requestMessage.MessageTypeCase, requestMessage.Id, requestMessage);

```

Comment:
I would try to extract methods for each case in the following switch, then continue that for next 2 levels of switches. The names and arguments of those methods would give us some insights when reading the code.